components/dashboard/src/workspaces/CreateWorkspacePage.tsx

```diff
@@ -0,0 +1,346 @@
+import { StartWorkspaceOptions } from "../start/start-workspace-options";
+import { StartWorkspaceError } from "../start/StartPage";
+import { useCurrentUser } from "../user-context";
+import { SelectAccountModal } from "../user-settings/SelectAccountModal";
+import Spinner from "../icons/Spinner.svg";
+import { useFeatureFlags } from "../contexts/FeatureFlagContext";
+
+export const useNewCreateWorkspacePage = () => {
+    const { startWithOptions } = useFeatureFlags();
+    const user = useCurrentUser();
+    return startWithOptions || user?.additionalData?.isMigratedToTeamOnlyAttribution;
+};
+
+export function CreateWorkspacePage() {
+    const user = useCurrentUser();
+    const location = useLocation();
+    const props = StartWorkspaceOptions.parseSearchParams(location.search);
+
+    const [useLatestIde, setUseLatestIde] = useState<boolean | undefined>(
+        props.ideSettings?.useLatestVersion !== undefined
+            ? props.ideSettings.useLatestVersion
+            : !!user?.additionalData?.ideSettings?.useLatestVersion,
+    );
+    const [selectedIde, setSelectedIde] = useState(
+        props.ideSettings?.defaultIde !== undefined
+            ? props.ideSettings.defaultIde
+            : user?.additionalData?.ideSettings?.defaultIde,
+    );
+    const [selectedWsClass, setSelectedWsClass] = useState<string | undefined>(props.workspaceClass);
+    const [errorWsClass, setErrorWsClass] = useState<string | undefined>(undefined);
+    const [repo, setRepo] = useState<string | undefined>(location.hash.substring(1));
+    const onSelectEditorChange = useCallback(
+        (ide: string, useLatest: boolean) => {
+            setSelectedIde(ide);
+            setUseLatestIde(useLatest);
+        },
+        [setSelectedIde, setUseLatestIde],
+    );
+    const [errorIde, setErrorIde] = useState<string | undefined>(undefined);
+    const [creating, setCreating] = useState<boolean>(false);
+
+    const [createResult, setCreateResult] = useState<WorkspaceCreationResult | undefined>(undefined);
+    const [createError, setCreateError] = useState<StartWorkspaceError | undefined>(undefined);
+    const [selectAccountError, setSelectAccountError] = useState<SelectAccountPayload | undefined>(undefined);
+
+    const createWorkspace = useCallback(
+        async (options?: Omit<GitpodServer.CreateWorkspaceOptions, "contextUrl">) => {
+            // add options from search params
+            const opts = options || {};
+
+            if (!opts.workspaceClass) {
+                opts.workspaceClass = selectedWsClass;
+            }
+            if (!opts.ideSettings) {
+                opts.ideSettings = {
+                    defaultIde: selectedIde,
+                    useLatestVersion: useLatestIde,
+                };
+            }
+            if (!repo) {
+                return;
+            }
+
+            try {
+                setCreating(true);
```

Comment:
With this being new code, it would be really nice to use a react query mutation for this. If you don't have time, maybe we could drop a comment, and I can follow up with moving it there. It'll help with adding consistency for managing things like the creating/saving/error states of async api calls like this.

There are a couple examples in the `src/data` directory if you end up wanting to give it a shot though. There's already a few workspace related queries/mutations present, so could add a new `data/workspaces/create-workspace-mutation.ts`. A pretty simple example for reference exists in `data/workspaces/stop-workspace-mutation.ts`. An extra mile could be invalidating the list workspaces query in an `onSuccess` handler for the new mutation. It'd look something like:

```ts
onSuccess: () => {
    const queryKey = getListWorkspacesQueryKey();
    queryClient.invalidateQueries({ queryKey });
}
```

Then it'd remove the need to manually manage a few pieces of state in this component, namely `creating`, `createError`, `createResult`.

Reply:
Not sure I understand for what you want to use a mutation. Can you clarify?

Reply:
Mutation is a term react-query uses for how it provides an abstraction for any api calls that mutate data, as opposed to a query, that just fetches data.

For example, [here's a](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/data/workspaces/stop-workspace-mutation.ts) "mutation" for the stop workspace api call. It's basically a hook wrapper around react-query's `useMutation()` hook.

[Here's what](https://github.com/gitpod-io/gitpod/blob/facc352acf6821c185553e2d8bb7ea4b859f1c17/components/dashboard/src/workspaces/WorkspaceOverflowMenu.tsx#L34-L50) it looks like when that mutation is used in a component.

In the case here, we actually want to await the mutation (rather than fire and forget), so it'd look slightly different, we'd use the `mutateAsync` fn like [here](https://github.com/gitpod-io/gitpod/blob/64d46a9aacb2e563ac436aa527f1bdb9df631a72/components/dashboard/src/teams/git-integrations/GitIntegrationListItem.tsx#L38-L45).

The "mutation" return value from calling the hook has some properties, like `isLoading`, `isError`, `data` (the returned result), and `error` if an error was thrown. These can be used in place of manually managing that state in the component [like here](https://github.com/gitpod-io/gitpod/blob/64d46a9aacb2e563ac436aa527f1bdb9df631a72/components/dashboard/src/teams/git-integrations/GitIntegrationListItem.tsx#L77-L78).

There's a lot more they can do that we can take advantage of over time. If you're curious, here's the full docs. Also, This would be something I would be happy to refactor as a followup, but wanted to mention it, as I think it'd be great if our new stuff could take advantage of react-query.

https://tanstack.com/query/v4/docs/react/guides/mutations

Reply:
Would be great if you could do that refactoring in a follow up. Thanks!